### PR TITLE
XLSX Importer

### PR DESCRIPTION
**Pull request by athityakumar**
Tests are yet to be beautified. Those changes are intended to be either here or, in PR #27 - depending on whichever PR is merged later.

### CODE REVIEW

File: lib/daru/io/importers/xlsx.rb [hunk 1]
```diff
@@ -0,0 +1,96 @@
+require 'daru/io/importers/base'
+
+module Daru
+  module IO
+    module Importers
+      class XLSX < Base
+        Daru::DataFrame.register_io_module :from_xlsx, self
+
```
Comment (athityakumar) [inline lib/daru/io/importers/xlsx.rb]: Is the XLSX Importer intended be used directly as `Daru::DataFrame.from_xlsx(...)`, or to be redirected from the Excel Importer if filename ends with `.xlsx`?

File: spec/fixtures/README.md [hunk 2]
```diff
@@ -6,6 +6,12 @@
 - `twitter.avro` : Downloaded from [here](https://github.com/miguno/avro-hadoop-starter/blob/master/src/test/resources/avro/twitter.avro). An AVRO schema with simple fields for Twitter users like `username`, `timestamp`, `tweet`.
 - `users.avro` : Downloaded from [here](https://github.com/apache/spark/blob/master/examples/src/main/resources/users.avro). A generic AVRO schema with fields like `name`, `favorite_color` and `favorite_numbers`.
 
+### Excelx
+
+- `LOBSTAHS_rt.windows.xlsx` : Downloaded from [here](https://github.com/vanmooylipidomics/LOBSTAHS/blob/master/inst/doc/xlsx/LOBSTAHS_rt.windows.xlsx). Contains data about Lipid and Oxylipin Biomarker Screening Through Adduct Hierarchy Sequences (LOBSTASHS). Conatins two sheets called `LOBSTAHS_rt.windows` and `Notes`.
+- `Microcode.xlsx` : Downloaded from [here](https://github.com/tkim371/CS2200/blob/master/proj2/project2/Microcode.xlsx).
+- `Stock-counts-sheet.xlsx` : Downloaded from [here](https://www.exact.com/uk/images/downloads/getting-started-excel-sheets/Stock-counts-sheet.xlsx). Contains data about stocks. Helps in ensuring that HTML tags of cell attributes are striped off, while constructing the DataFrame.
+
```
Comment (athityakumar) [inline spec/fixtures/README.md]: I somehow completely missed to add xlsx fixture files, during PR #17. Sorry! :-/

File: lib/daru/io/importers/xlsx.rb [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+          @path    = path
+          @sheet   = sheet
+          @headers = headers
+        end
+
+        def call
+          book      = Roo::Excelx.new(@path)
+          worksheet = book.sheet(@sheet)
+          data      = worksheet.to_a
+          data      = strip_html_tags(data)
```
Comment (zverok) [inline lib/daru/io/importers/xlsx.rb]: Why is this necessary? Is there a lot of examples in the wild with HTML tags in cells?..

File: lib/daru/io/importers/xlsx.rb [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            row.map do |ele|
+              next ele unless ele.is_a?(String)
+              ele.gsub(/<[^>]+>/, '')
+            end
+          end
+        end
+      end
+    end
+  end
+end
```
Comment (zverok) [inline lib/daru/io/importers/xlsx.rb]: Questions to the impl:
* what about skipping some cols and rows? I believe I've seen a lot of governmental open data, for ex., with large headers above the data itself;
* a lot of Excel file I've seen also had "index" (e.g. row names) in them, what about it?
* it is high probability of merged cells, especially in index/order, WDYT of it?
* a lot of Excel files have formulae in cells, WDYT of it?

File: spec/daru/io/importers/xlsx_spec.rb [hunk 5]
```diff
@@ -0,0 +1,67 @@
+RSpec.describe Daru::IO::Importers::XLSX do
+  subject { described_class.new(path, sheet: sheet, headers: headers).call }
+
+  let(:sheet)   { 0    }
+  let(:headers) { true }
+
+  context 'parses default sheet' do
```
Comment (zverok) [inline spec/daru/io/importers/xlsx_spec.rb]: Why "default" is 0?..

File: spec/daru/io/importers/xlsx_spec.rb [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+      )
+    end
+  end
+
+  before do
+    %w[LOBSTAHS_rt.windows Microcode Stock-counts-sheet].each do |file|
+      WebMock
+        .stub_request(:get,"http://dummy-remote-url/#{file}.xlsx")
+        .to_return(status: 200, body: File.read("spec/fixtures/excelx/#{file}.xlsx"))
+      WebMock.disable_net_connect!(allow: /dummy-remote-url/)
```
Comment (zverok) [inline spec/daru/io/importers/xlsx_spec.rb]: Is it needed?.. I believe we can just disable_net_connect in spec_helper.rb (or nowhere, it is done by default), without those dances.

File: spec/daru/io/importers/xlsx_spec.rb [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+  context 'parses nil elements in sheet' do
+    let(:path)  { 'spec/fixtures/excelx/Stock-counts-sheet.xlsx' }
+    let(:sheet) { 2 }
+
+    it                              { is_expected.to be_an(Daru::DataFrame)    }
+    its(:ncols)                     { is_expected.to eq(7)                     }
+    its(:nrows)                     { is_expected.to eq(15)                    }
+    its(:index)                     { is_expected.to eq((0..14).to_a.to_index) }
+    its('Item code.first')          { is_expected.to eq(nil)                   }
```
Comment (zverok) [inline spec/daru/io/importers/xlsx_spec.rb]: What it was in source file? Empty cell? Then let's name the context this way. Like `context "when sheet contains empty cells"`.
And the second question: how do you decide that it should be `nil`, and not, say, `""`?

Comment (zverok) [inline spec/daru/io/importers/xlsx_spec.rb]: BTW, on context names. There are several agreements, but generally we follow this structure:
```ruby
describe Class do
  describe '#method_name' do
    context 'in this circumstances' do # e.g. "when this and that passed", "on erroneous request", "if requested from network"
    end
  end
end
```

E.g., context description should .... well, you know, _describe context_ of this group of tests :)

File: lib/daru/io/importers/xlsx.rb (hunk 3 above)
Comment (zverok) [inline lib/daru/io/importers/xlsx.rb]: Oh damn. Roo converts any formatting to HTML unconditionally :( What could possibly lead any sane developer to this decision :rage:

File: spec/daru/io/importers/xlsx_spec.rb (hunk 5 above)
Comment (athityakumar) [inline spec/daru/io/importers/xlsx_spec.rb]: Wouldn't it be logical to parse the 1st sheet by default? The default is set as 0, even for the `from_excel` method.

Comment (zverok) [inline spec/daru/io/importers/xlsx_spec.rb]: Yes, it is logical. What I am concerned about here is by concept description, you are testing "what it would be if there's no sheet param passed" (I am guessing it by "default" word). But in fact, you are testing the `sheet: 0` explicitly passed.

File: lib/daru/io/importers/xlsx.rb (hunk 4 above)
Comment (athityakumar) [inline lib/daru/io/importers/xlsx.rb]: I'll soon add the `:skiprows` and `:skipcols` options. They seem to be quite necessary for an initial draft.

Regarding the rest of the features, let me open issue #36 to make sure these suggestions don't get missed. 😄

File: lib/daru/io/importers/excel.rb [hunk 8]
```diff
@@ -4,19 +4,29 @@ module Daru
   module IO
     module Importers
       class Excel < Base
-        Daru::DataFrame.register_io_module :from_excel, self
+        Daru::DataFrame.register_io_module :from_excel do |*args|
+          if args.first.end_with? '.xlsx'
+            require 'daru/io/importers/excelx'
```
Comment (v0dro) [inline lib/daru/io/importers/excel.rb]: Does this mean that a `require` will be called each time the `from_excel` method is called? That's inefficient. You should ideally perform all the requires when the gem is required or when a specific importer/exporter is required.

File: lib/daru/io/importers/excel.rb [hunk 9]
```diff
@@ -45,22 +55,26 @@ class Excel < Base
         #   #    3        4    Franz      nil    Paris      nil
         #   #    4        5   George      5.5     Tome    a,b,c
         #   #    5        6  Fernand      nil      nil      nil
-        def initialize(path, worksheet_id: 0)
+        def initialize(path, sheet: 0, headers: true)
           optional_gem 'spreadsheet', '~> 1.1.1'
```
Comment (v0dro) [inline lib/daru/io/importers/excel.rb]: Is it possible to have the `optional_gem` method run at the time of the gem require? Or how about "trying" to require all gems at gem require first and then throw warnings only when the specific method is called?

For example, the `spreadsheet` gem will be required when the user calls `require 'daru/io'` or `require 'daro/io/importers/excel'`. This will not throw any warnings. However, when the user calls `from_excel` you can throw the error. You check whether a library is present or not with this: https://stackoverflow.com/questions/8424769/how-do-you-check-if-a-library-ruby-gem-has-been-loaded

@zverok  thoughts?

Comment (v0dro) [inline lib/daru/io/importers/excel.rb]: @athityakumar also, can you add documentation to the `optional_gem` method?

https://github.com/athityakumar/daru-io/blob/master/lib/daru/io/base.rb#L7

File: lib/daru/io/importers/excel.rb [hunk 10]
```diff
@@ -45,22 +55,26 @@ class Excel < Base
         #   #    3        4    Franz      nil    Paris      nil
         #   #    4        5   George      5.5     Tome    a,b,c
         #   #    5        6  Fernand      nil      nil      nil
-        def initialize(path, worksheet_id: 0)
+        def initialize(path, sheet: 0, headers: true)
```
Comment (v0dro) [inline lib/daru/io/importers/excel.rb]: Any specific reason for changing to `sheet`? It will break backwards compatibility.

File: lib/daru/io/importers/excelx.rb [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+        #   #=> <Daru::DataFrame(16x7)>
+        #                   0           1          2          3          4          5        6
+        #        0      Status Stock coun  Item code        New Descriptio Stock coun Offset G/L
+        #        1          H          1        nil        nil New stock  2014-08-01        nil
+        #        2        nil          1  IND300654          2 New stock  2014-08-01      51035
+        #        3        nil          1   IND43201          5 New stock  2014-08-01      51035
+        #        4        nil          1   OUT30045          3 New stock  2014-08-01      51035
+        #       ...       ...        ...     ...           ...     ...       ...           ...
+        #
+        # @example Importing from a remote URL
```
Comment (v0dro) [inline lib/daru/io/importers/excelx.rb]: How will this documentation be easily accessible to someone who is using the `from_excel` method? This implementation of using importer and exporter classes is opaque to the user, right?

File: lib/daru/io/importers/excel.rb (hunk 10 above)
Comment (athityakumar) [inline lib/daru/io/importers/excel.rb]: For the excelx (xlsx) importer, name of the keyword argument has been chosen as `sheet`, as it could even be a `String` and `worksheet_id` seems a bit misleading to only `Integer` type (Of course, the documentation should take care of it). And to maintain consistency between excel and excelx importer, `worksheet_id` was renamed to `sheet`.

But this was prior to commit f9a8356, where excel importer redirected to excelx importer and both methods were forced to have same argument names. Now that redirect is done by `register_io_module`, this can certainly be reverted back to `worksheet_id`.

File: lib/daru/io/importers/excelx.rb (hunk 11 above)
Comment (athityakumar) [inline lib/daru/io/importers/excelx.rb]: As the linkages are done dynamically, YARD won't be able to recognize the linkages. The simplest workaround I can think for this, is to include their respective hyperlinks to the `Daru::DataFrame#register_io_module` YARD comments. Like,

```ruby
module Daru
  class DataFrame  
    class << self
      # Links +Daru::DataFrame+ Import / Export methods to corresponding 
      # +Daru::IO+ Importer / Exporter classes.
      #
      # - {Daru::IO::Importers::CSV#initialize #from_csv method}
      # - {Daru::IO::Importers::Redis#initialize #from_redis method}
      # ...
      def register_io_module(function, instance)
          ...
      end
    end
  end
end
```

I think it's reasonable to expect users to open the DataFrame page to have a look at the Importers & Exporters, and find a list of them as shown below.
image
Does this suffice?

File: lib/daru/io/importers/excel.rb (hunk 8 above)
Comment (athityakumar) [inline lib/daru/io/importers/excel.rb]: Yes, a `require` will be called each time the `from_excel` method is called with a `.xlsx` file. But I'm not sure why it'd be inefficient - isn't `require` by default done only once?  And would this make it better?
```ruby
require 'daru/io/importers/excelx' unless defined?(Daru::IO::Importers::Excelx)
```

> perform all the requires when the gem is required or when a specific importer/exporter is required.

When a user's use-case is only `.xls` files, the excelx importer isn't needed to be `require`d right? (Or is unconditionally requiring excelx importer when requiring excel importer, a better standard?)

File: lib/daru/io/importers/excel.rb (hunk 9 above)
Comment (athityakumar) [inline lib/daru/io/importers/excel.rb]: @v0dro - Sure, will document the `#optional_gem` method. Meanwhile, intention of having `optional_gem` run at `#initialize` is same as that of conditionally requiring `excelx` : require only when a particular IO module is used. Sure, `optional_gem` can be tweaked to try requiring dependencies when an Importer is required and raise errors (`LoadError`, etc.) when the importer is called.

File: lib/daru/io/importers/excel.rb (hunk 8 above)
Comment (zverok) [inline lib/daru/io/importers/excel.rb]: > That's inefficient. 

Not very. `require` works only once, the next time Ruby's enternal mechanizms will just check whether require'd file is loaded already.

File: lib/daru/io/importers/excel.rb (hunk 9 above)
Comment (zverok) [inline lib/daru/io/importers/excel.rb]: > Is it possible to have the optional_gem method run at the time of the gem require? Or how about "trying" to require all gems at gem require first and then throw warnings only when the specific method is called?

@v0dro To be honest, I am strictly against it. `optional_gem` is not effect-less, it outputs "Hey, we need this and that gem to do what you want". Imagine daru-io v.2.3.4, which somebody wants to use to just import some CSV? It will output two full screens of `please install <this irrelevant stuff>` just for nothing.

I believe, calling `optional_gem` _exactly_ at the moment when somebody tries to really import/export `xlsx` is a good idea.

Side note: In future, there can exist code like 
```ruby
format = controller.params[:format]
dataframe = Daru.from(format, controller.params[:input])
```
...which could probably fail too late (e.g. works for 1000 users, and fails for 1001st, because format was `dbf`, and dependency is absent), but it is another story (about APIs where `format` is a variable, and then about APIs of "early check" that we support this and that format, and probably explicitly forbid those and those).

File: lib/daru/io/importers/excelx.rb (hunk 11 above)
Comment (zverok) [inline lib/daru/io/importers/excelx.rb]: @v0dro I believe that `daru-io` should think about it as a complementary yet separate gem. 
It is a necessary compromise we've already discussed with Athitya: whether we document exact modules that do the work (`Daru::IO::Whatever`) or their generated methods? (`DataFrame.from_whatever`)
My answer is the former. 
Let's think about somebody who uses daru-io as a foundation to add another format, say, FITS (it is astronomy data).
What they do?
1. `require 'daru-io'`
2. Create `Daru::IO::FITS` class by a simple straightforward instructions, and document it.
3. Call `Daru::IO.register_io_module`.

Of course, we **can** ask them to implement pseudo-docs for `from_fits`, YARD allows this... But do we really faithfully **want** to?..

So, I believe `Daru::IO` docs should be saying something like... "Blah, blah, when you include this and that, you receive a ton of `Daru::DataFrame.from_this_and_that`, please refer to `Daru::IO::ThisAndThatImporter` for docs".

It is one option. Another one is stop polluting DF with `from_foo`/`to_bar` methods, and define for `Daru::IO` API like `Daru::IO::FromExcel.new('path').call` / `Daru::IO::ToExcel.new(df).call('path')`.

File: lib/daru/io/importers/excelx.rb [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+        #
+        #   #=> <Daru::DataFrame(15x7)>
+        #              Status Stock coun  Item code        New Descriptio Stock coun Offset G/L
+        #        0          H          1        nil        nil New stock  2014-08-01        nil
+        #        1        nil          1  IND300654          2 New stock  2014-08-01      51035
+        #        2        nil          1   IND43201          5 New stock  2014-08-01      51035
+        #        3        nil          1   OUT30045          3 New stock  2014-08-01      51035
+        #       ...       ...        ...     ...           ...     ...       ...           ...
+        def initialize(path, sheet: 0, order: true, index: false, skiprows: 0, skipcols: 0)
+          optional_gem 'roo', '~> 2.7.0'
```
Comment (zverok) [inline lib/daru/io/importers/excelx.rb]: In fact, this can be metaprogrammed. Like this:

```ruby
class Excelx < Base
  requires_gem 'roo', '~> 2.7.0' # says that on real creation of object the check should be done
```

Another thought is you probably should not specify dependencies version so narrow for general-usage gem (which leads to idea of testing with multiple dependencies versions... So we should postpone this, just don't forget)

File: lib/daru/io/link.rb [hunk 13]
```diff
@@ -1,7 +1,9 @@
 module Daru
   class DataFrame
     class << self
-      def register_io_module(function, instance)
+      def register_io_module(function, instance=nil)
+        return define_singleton_method(function) { |*args| yield(*args) } if block_given?
```
Comment (zverok) [inline lib/daru/io/link.rb]: With explicit `&block` in args, you could've just `define_singleton_method(function, &block)` here.

File: lib/daru/io/importers/excelx.rb (hunk 12 above)
Comment (athityakumar) [inline lib/daru/io/importers/excelx.rb]: Acknowledged. Keeping #25 open and also opened #41 for future references. 👍

File: lib/daru/io/importers/excel.rb (hunk 9 above)
Comment (v0dro) [inline lib/daru/io/importers/excel.rb]: Ok makes sense to call `optional_gem` only when the method is called.

File: lib/daru/io/importers/excelx.rb (hunk 11 above)
Comment (v0dro) [inline lib/daru/io/importers/excelx.rb]: Lets just stick to referring users to the relevant module in that case. Changing the API would not look good.

**Comment by v0dro** [comment]
Approved. Merge?

**Comment by athityakumar** [comment]
@v0dro - Sure, merging this in. 😃
